File: page-storage-system/storage/buffer_pool.py
from dataclasses import dataclass
from typing import Dict

from .constants import PAGE_SIZE
from .errors import StorageError
from .page_manager import PageManager
from .replacement import make_policy


@dataclass
class BufferFrame:
    data: bytearray
    dirty: bool = False


class BufferPool:
    def __init__(self, page_manager: PageManager, capacity: int = 16, policy: str = "LRU", log: bool = True):
        if not isinstance(capacity, int) or isinstance(capacity, bool) or capacity <= 0:
            raise StorageError("BUFFER_INVALID_CAPACITY", "capacity 必须是正整数")
        self.page_manager = page_manager
        self.capacity = capacity
        self.policy = make_policy(policy)
        self.log_enabled = bool(log)
        self.frames: Dict[int, BufferFrame] = {}
        self.hits = 0
        self.misses = 0
        self.evictions = 0
        self.flushes = 0
        self.events = []

    def _log(self, event: str, page_id: int, victim_page_id=None):
        if self.log_enabled:
            self.events.append({"event": event, "pageId": page_id, "victimPageId": victim_page_id})

    def _evict_if_needed(self):
        if len(self.frames) < self.capacity:
            return None, None
        victim = self.policy.choose_victim(self.frames.keys())
        frame = self.frames[victim]
        flushed = None
        if frame.dirty:
            self.page_manager.write_page(victim, bytes(frame.data))
            frame.dirty = False
            self.flushes += 1
            flushed = victim
            self._log("FLUSH", victim)
        del self.frames[victim]
        self.policy.record_remove(victim)
        self.evictions += 1
        self._log("EVICT", victim, victim_page_id=victim)
        return victim, flushed

    def get_page(self, page_id: int):
        if page_id in self.frames:
            self.hits += 1
            self.policy.record_access(page_id)
            self._log("HIT", page_id)
            frame = self.frames[page_id]
            return bytes(frame.data), True, frame.dirty

        self.misses += 1
        self._log("MISS", page_id)
        self._evict_if_needed()
        data = self.page_manager.read_page(page_id)
        self.frames[page_id] = BufferFrame(bytearray(data), False)
        self.policy.record_insert(page_id)
        return bytes(data), False, False

    def put_page(self, page_id: int, data: bytes, dirty: bool = True):
        data = bytes(data)
        if len(data) != PAGE_SIZE:
            raise StorageError("INVALID_PAGE_SIZE", "写入页必须恰好为 4096 字节", page_id=page_id)
        if not self.page_manager.is_allocated(page_id):
            raise StorageError("PAGE_NOT_ALLOCATED", "页不存在或已经释放", page_id=page_id)

        evicted = flushed = None
        if page_id not in self.frames:
            evicted, flushed = self._evict_if_needed()
            self.frames[page_id] = BufferFrame(bytearray(data), bool(dirty))
            self.policy.record_insert(page_id)
        else:
            frame = self.frames[page_id]
            frame.data[:] = data
            frame.dirty = frame.dirty or bool(dirty)
            self.policy.record_access(page_id)
        return evicted, flushed

    def flush_page(self, page_id: int) -> bool:
        if page_id not in self.frames:
            raise StorageError("BUFFER_PAGE_NOT_RESIDENT", "页当前不在 Buffer Pool 中", page_id=page_id)
        frame = self.frames[page_id]
        if frame.dirty:
            self.page_manager.write_page(page_id, bytes(frame.data))
            frame.dirty = False
            self.flushes += 1
            self._log("FLUSH", page_id)
        return True

    def flush_all(self):
        flushed = []
        for page_id in list(self.frames.keys()):
            frame = self.frames[page_id]
            if frame.dirty:
                self.page_manager.write_page(page_id, bytes(frame.data))
                frame.dirty = False
                self.flushes += 1
                self._log("FLUSH", page_id)
                flushed.append(page_id)
        self.page_manager.sync()
        return flushed

    def discard_page(self, page_id: int):
        self.frames.pop(page_id, None)
        self.policy.record_remove(page_id)

    def stats(self):
        return {
            "capacity": self.capacity,
            "size": len(self.frames),
            "hits": self.hits,
            "misses": self.misses,
            "evictions": self.evictions,
            "flushes": self.flushes,
            "policy": self.policy.name,
        }

File: page-storage-system/src/main/java/org/chainpage/storage/BufferPool.java
package org.chainpage.storage;

import java.util.*;

public final class BufferPool {
    private static final class Frame {byte[] data;long generation;boolean dirty;final List<Long> lsns=new ArrayList<>();Frame(byte[]d,long g,boolean x){data=d;generation=g;dirty=x;}}
    private final PageManager pages;private final WalManager wal;private final LockManager locks;private final int capacity;private ReplacementPolicy policy;
    private final Map<Integer,Frame> frames=new HashMap<>();private final List<Map<String,Object>> events=new ArrayList<>();private long hits,misses,evictions,flushes;
    BufferPool(PageManager p,WalManager w,LockManager l,int capacity,String policy){if(capacity<=0)throw new StorageException("BUFFER_INVALID_CAPACITY","capacity 必须为正整数");pages=p;wal=w;locks=l;this.capacity=capacity;this.policy=ReplacementPolicy.make(policy);pages.setPageInvalidator(this::beforeFree);}
    private String owner(String owner){return owner==null||owner.isBlank()?"buffer-"+Thread.currentThread().getId():owner;}
    public synchronized Page getPage(int id,String owner){pages.requireAllocated(id);try(LockManager.Lease ignored=locks.acquire(id,"READ",owner(owner))){Frame f=frames.get(id);if(f!=null&&f.generation!=pages.generation(id)){frames.remove(id);policy.remove(id);f=null;}if(f!=null){hits++;policy.access(id);event("HIT",id,null);return new Page(f.data.clone(),true,f.dirty);}misses++;event("MISS",id,null);byte[] data=pages.readPage(id);evict();frames.put(id,new Frame(data,pages.generation(id),false));policy.insert(id);return new Page(data.clone(),false,false);}}
    public synchronized Put putPage(int id,byte[] data,boolean dirty,String owner,int tx){FileManager.page(data,id);pages.requireAllocated(id);try(LockManager.Lease ignored=locks.acquire(id,"WRITE",owner(owner))){Frame f=frames.get(id);if(f!=null&&f.generation!=pages.generation(id)){frames.remove(id);policy.remove(id);f=null;}byte[] before=f==null?pages.readPage(id):f.data.clone();Integer victim=null,flushed=null;if(f==null){int[] e=evict();victim=e[0]<0?null:e[0];flushed=e[1]<0?null:e[1];f=new Frame(data.clone(),pages.generation(id),dirty);frames.put(id,f);policy.insert(id);}else{f.data=data.clone();f.dirty|=dirty;policy.access(id);}if(f.dirty&&!Arrays.equals(before,data))f.lsns.add(wal.append(tx,id,before,data));return new Put(victim,flushed);}}
    private int[] evict(){if(frames.size()<capacity)return new int[]{-1,-1};int id=policy.victim(frames.keySet());Frame f=frames.get(id);try(LockManager.Lease ignored=locks.acquire(id,"WRITE",owner(null))){boolean wrote=flushFrame(id,f);frames.remove(id);policy.remove(id);evictions++;event("EVICT",id,id);return new int[]{id,wrote?id:-1};}}
    public synchronized boolean flushPage(int id,String owner){try(LockManager.Lease ignored=locks.acquire(id,"WRITE",owner(owner))){Frame f=frames.get(id);if(f==null)throw new StorageException("BUFFER_PAGE_NOT_RESIDENT","页不在 Buffer Pool",id);flushFrame(id,f);return true;}}
    public synchronized List<Integer> flushAll(){List<Integer> out=new ArrayList<>();for(int id:new ArrayList<>(frames.keySet())){try(LockManager.Lease ignored=locks.acquire(id,"WRITE",owner(null))){Frame f=frames.get(id);if(f!=null&&flushFrame(id,f))out.add(id);}}pages.sync();return out;}
    private boolean flushFrame(int id,Frame f){if(!f.dirty)return false;if(!pages.isAllocated(id)||f.generation!=pages.generation(id)){f.dirty=false;f.lsns.clear();return false;}pages.writePage(id,f.data);pages.sync();for(long l:f.lsns)wal.applied(l);f.lsns.clear();f.dirty=false;flushes++;event("FLUSH",id,null);return true;}
    private void beforeFree(int id){try(LockManager.Lease ignored=locks.acquire(id,"WRITE",owner(null))){Frame f=frames.get(id);if(f!=null){flushFrame(id,f);frames.remove(id);policy.remove(id);}}}
    synchronized void discard(int id){Frame f=frames.get(id);if(f!=null&&f.dirty)throw new StorageException("BUFFER_DIRTY_DISCARD","不能丢弃脏页",id);frames.remove(id);policy.remove(id);}
    synchronized void forceDiscard(int id){frames.remove(id);policy.remove(id);}
    synchronized void clear(){frames.clear();policy=ReplacementPolicy.make(policy.name());}
    synchronized Map<String,Object> stats(){Map<String,Object> m=JsonFiles.map();m.put("capacity",capacity);m.put("size",frames.size());m.put("hits",hits);m.put("misses",misses);m.put("evictions",evictions);m.put("flushes",flushes);m.put("policy",policy.name());return m;}
    synchronized Map<String,Object> setPolicy(String name){ReplacementPolicy p=ReplacementPolicy.make(name);frames.keySet().forEach(p::insert);policy=p;return Map.of("policy",p.name());}
    synchronized List<Map<String,Object>> events(boolean clear){List<Map<String,Object>> out=events.stream().map(LinkedHashMap::new).map(x->(Map<String,Object>)x).toList();if(clear)events.clear();return out;}
    private void event(String kind,int id,Integer victim){Map<String,Object> e=JsonFiles.map();e.put("event",kind);e.put("pageId",id);e.put("victimPageId",victim);events.add(e);}
    public record Page(byte[] data,boolean hit,boolean dirty){} public record Put(Integer evicted,Integer flushedPageId){}
}
